### ISSUE TITLE
Add `from_handle` constructors for various Vulkan object types.

### ISSUE THREAD

**Issue by Rua** [vulkano-rs/vulkano#1935]
On Discord, a wish was expressed for the ability to create Vulkano types from raw Vulkan object handles (whose types are defined by Ash). Although there are many object types to implement this for, I think this would be relatively easy for new contributors to implement. Some implementation guidelines:
- The new constructor, I'm naming it `from_handle` for the moment, should be `unsafe`.
- Whereas the normal `new` constructor usually takes a `Device` and some kind of `Info` struct, the `from_handle` constructor would take a `Device`, a raw handle of the appropriate Ash type, and the `Info` struct.
- The function must have a `# Safety` header explaining that the handle must be a valid Vulkan object handle, and that the parameters in the given `Info` struct must match those that were previously used to create the handle.
- Because the handle is assumed to be already valid, little to no validation will be needed inside `from_handle`. The data in the `Info` struct could probably just be moved into the Vulkano type without any checks.

If anyone wants to try implementing this, but is unsure about the number of changes required, you can start by implementing it only for a single, not too complex type (e.g. `Sampler`, `QueryPool`).

**Comment by p3t3rix** [comment]
So i tried my hand at the QueryPool by liberally copying parts of the ctor, something along those lines:
```
/// Creates a new `QueryPool` from an ash-handle
/// # Safety
/// The `handle` has to be a valid vulkan object handle and 
/// the `create_info` must match the info used to create said object
pub unsafe fn from_handle(device: Arc<Device>,
                          create_info: QueryPoolCreateInfo,
                          handle : ash::vk::QueryPool
) -> Result<Arc<QueryPool>, QueryPoolCreationError>{
    let QueryPoolCreateInfo {
        query_type,
        query_count,
        _ne: _,
    } = create_info;
    return Ok(Arc::new(QueryPool {
        handle,
        device,
        query_type,
        query_count,
    }));
}
```
- Is this the right direction or am i missing something ? 
- Also should there be tests for these and if so how should they look like ?
- Should this even return an error as the object is already created ?

**Comment by Rua** [comment]
Yes, I think that's the right direction. Personally I think the `handle` parameter should come before `create_info`, but that's just bikeshedding and I can't say why it seems better. There shouldn't be an error return, because there are no errors to report; like you say, the object is already created.

I don't think tests are necessary, because there isn't really anything to test, as far as I can think of.

**Comment by p3t3rix** [comment]
After https://github.com/vulkano-rs/vulkano/pull/1938 the following things wrap ash internally and still need to be discussed if they need a from_handle function and how it should look like:
-   ShaderModule
-   GraphicsPipeline
-   ComputePipeline
-   PipelineCache (not needed as far as i am concerned)
-   DeviceMemory
-   Instance
-   Device
-   UnsafeDescriptorSet (has it already under new() but not public )
-   UnsafeCommandPoolAlloc
-   UnsafeCommandBuffer

**Comment by Rua** [comment]
I think `DeviceMemory` could have a `from_handle` relatively straightforwardly. `Instance` and `Device` may not be too difficult either, but I haven't looked at what blockers there may be.

**Comment by yshui** [comment]
hey, i think what #1396 was asking for are all there yet. we also need `from_handle` for `VkDevice` and `VkInstance`.

**Comment by Rua** [comment]
That's already done.

**Comment by yshui** [comment]
@Rua it is? i don't see a `Device::from_handle`: https://docs.rs/vulkano/latest/vulkano/device/struct.Device.html

is it not released yet?

**Comment by digama0** [comment]
I also ran into issues when integrating with OpenXR. Not as many as some of the other posters (I was able to construct a `Instance`, `PhysicalDevice`, `Device`) but I got stuck when [generating a `Swapchain`](https://github.com/Ralith/openxrs/blob/d05c74a9b296d8f29022ca5e943cb8dafd4a6735/openxr/examples/vulkan.rs#L524-L540) because this generates an `xr::Swapchain` (aka [XrSwapchain](https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#XrSwapchain)), not a `vk::Swapchain` (aka [VK_KHR_swapchain](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_swapchain.html)). On its own this is okay, but when we [pull out some `vk::Image`s from it](https://github.com/Ralith/openxrs/blob/d05c74a9b296d8f29022ca5e943cb8dafd4a6735/openxr/examples/vulkan.rs#L551) there is no provenance data that I can use to populate the arguments of `Image::from_raw` or `Image::from_swapchain` or even `RawImage::from_handle`. And without those `Image`s I can't create a `Framebuffer` or run the pipeline later, because all of these types expect other vulkano types and I can't just pass raw values to them.
